src/Gui/OverlayManager.cpp

```diff
@@ -0,0 +1,2061 @@
+                if (o->tabWidget == it->second) {
+                    o->addWidget(dw, true);
+                    d->onToggleDockWidget(dw, 3);
+                    break;
+                }
+            }
+            d->refresh();
+        }
+    }
+}
+
+void OverlayManager::setupDockWidget(QDockWidget *dw, int dockArea)
+{
+    (void)dockArea;
+    d->setupTitleBar(dw);
+}
+
+void OverlayManager::unsetupDockWidget(QDockWidget *dw)
+{
+    d->toggleOverlay(dw, ToggleMode::Unset);
+}
+
+void OverlayManager::onToggleDockWidget(bool checked)
+{
+    auto action = qobject_cast<QAction*>(sender());
+    if(!action)
+        return;
+    d->onToggleDockWidget(qobject_cast<QDockWidget*>(action->parent()), checked);
+}
+
+void OverlayManager::onDockVisibleChange(bool visible)
+{
+    auto dock = qobject_cast<QDockWidget*>(sender());
+    if(!dock)
+        return;
+    FC_TRACE("dock " << dock->objectName().toUtf8().constData()
+            << " visible change " << visible << ", " << dock->isVisible());
+}
+
+void OverlayManager::onTaskViewUpdate()
+{
+    auto taskview = qobject_cast<TaskView::TaskView*>(sender());
+    if (!taskview)
+        return;
+    QDockWidget *dock = nullptr;
+    DockWnd::ComboView *comboview = nullptr;
+    for (QWidget *w=taskview; w; w=w->parentWidget()) {
+        if ((dock = qobject_cast<QDockWidget*>(w)))
+            break;
+        if (!comboview)
+            comboview = qobject_cast<DockWnd::ComboView*>(w);
+    }
+    if (dock) {
+        if (comboview && comboview->hasTreeView()) {
+            refresh();
+            return;
+        }
+        auto it = d->_overlayMap.find(dock);
+        if (it == d->_overlayMap.end()
+                || it->second->tabWidget->count() < 2
+                || it->second->tabWidget->getAutoMode() != OverlayTabWidget::AutoMode::TaskShow)
+            return;
+        d->onToggleDockWidget(dock, taskview->isEmpty() ? -2 : 2);
+    }
+}
+
+void OverlayManager::onDockWidgetTitleChange(const QString &title)
+{
+    if (title.isEmpty())
+        return;
+    auto widget = qobject_cast<QWidget*>(sender());
+    QDockWidget *dock = nullptr;
+    for (QWidget *w=widget; w; w=w->parentWidget()) {
+        if ((dock = qobject_cast<QDockWidget*>(w)))
+            break;
+    }
+    if(!dock)
+        return;
+    auto tabWidget = findTabWidget(dock);
+    if (!tabWidget)
+        return;
+    int index = tabWidget->dockWidgetIndex(dock);
+    if (index >= 0)
+        tabWidget->setTabText(index, title);
+}
+
+void OverlayManager::retranslate()
+{
+    d->retranslate();
+}
+
+void OverlayManager::refreshIcons()
+{
+    d->refreshIcons();
+}
+
+void OverlayManager::reload(ReloadMode mode)
+{
+    d->reload(mode);
+}
+
+void OverlayManager::raiseAll()
+{
+    d->raiseAll();
+}
+
+static inline bool
+isNear(const QPoint &a, const QPoint &b, int tol = 16)
+{
+    QPoint d = a - b;
+    return d.x()*d.x() + d.y()*d.y() < tol;
+}
+
+bool OverlayManager::eventFilter(QObject *o, QEvent *ev)

```

Comment:
Another long function: the easy refactor here is to split each case statement's content into its own function call.